### DISCUSSION TITLE
Using mirrors on an isolated machine without ca-certificates

### DISCUSSION THREAD

**Discussion by thom7660** [spack/spack#33492] [Q&A]
I am currently trying to install a package via spack on a machine that is isolated from the outside world via spack's "mirror" functionality. I can create the mirror and transfer it to the isolated machine no problem. However, when I try to install the package spack always fails to install the ca-certificates-mozilla package, even though it is included in the mirror, which then causes the entire installation to fail. The error I get is as shown:

```
No binary for ca-certificates-mozilla-2022-07-19-zgbpzgptdgyrfeiirtv3ejga2rhgii44 found: installing from source
==> Error: FetchError: All fetchers failed for spack-stage-ca-certificates-mozilla-2022-07-19-zgbpzgptdgyrfeiirtv3ejga2rhgii44  
```

I assume this is because the certificates are a dependency required by spack itself and should already be present on my system, which they are not. Is there any way to tell spack to ignore these certificates and continue with the installation? 

So far I have tried using the -k flag, as well as adding a second link in the ca-certifcates-mozzilla mirror directory that includes the string after 2022-07-19, both to no avail. Looking at the --debug output simply shows the binary not being found and all connection attempts timing out (naturally, since the machine can't access the internet). Perhaps spack is simply not usable in my case, but I wanted to see if there were any workarounds before I commit to building from source. I can provide any other information as needed and appreciate any insights!

**Comment by haampie** [comment]
Likely the issue is that `ca-certificates-mozilla` is a build-type dependency, which is not pushed to the mirror. But if you do `spack install --use-buildcache only` it should drop build-only type dependencies (and their dependencies).

**Reply by thom7660** [reply]
Thank you for the suggestion! I apologize for the delay, I wanted to look into spack's buildcache functionality before getting back to you. Unfortunately, I am running into the same issue even when including the suggested --use-buildcache only flag. Do I need to create a buildcache via the 'spack buildache' command and add it to the mirror before putting it on the isolated machine? Currently, I have only used the mirror command and have not explicitly created any caches other than the default '_source-cache'. The commands I used are as follows (for installing the 'precice' package):

```bash
$ spack mirror create -d precice_mirror -D precice@2.5.0
$ ssh isolated_machine
$ cd precice_mirror
$ spack mirror add local_filesystem file:///precice_mirror
$ spack env create -d .
$ spacktivate
$ spack add precice
$ spack install --use-buildcache only
```

The mirror is also set to be used first in my mirrors.yaml file.

I did take a quick stab at creating a buildcache and adding it as it's own mirror, but for some reason, I was unable to add the cache to the buildcache list on the isolated machine. I suspect this may be because the installation I used to build the cache uses dependencies that are different than the machine I am trying to install on.

I apologize if these are silly questions, I am very new to spack and am still trying to wrap my head around a lot of its functionality
